dsgrid/config/representative_period_time_dimension_config.py

```diff
@@ -44,10 +48,86 @@ def check_dataset_time_consistency(self, load_data_df):
+        df_time = spark.createDataFrame(model_time, schema=schema)
+
+        return df_time
+
+    def convert_dataframe(self, df=None, project_time_dim=None):
         # TODO: Create a dataframe with timestamps covering project_time_dim.model.ranges
         # for all model_years in the df, then join the df to that dataframe on the time columns.
         # Account for time zone.
+
+        if project_time_dim is None:
+            return df
+        if (
+            project_time_dim.list_expected_dataset_timestamps()
+            == self.list_expected_dataset_timestamps()
+        ):
+            return df
+
+        time_cols = self.get_timestamp_load_data_columns()
+        ptime_col = project_time_dim.get_timestamp_load_data_columns()
+        assert len(ptime_col) == 1, ptime_col
+        ptime_col = ptime_col[0]
+
+        # get unique timezones
+        geo_tz_values = [row.time_zone for row in df.select("time_zone").distinct().collect()]
+        geo_tz_names = [get_timezone(tz).tz_name for tz in geo_tz_values]
+
+        # create time map
+        project_time_df = project_time_dim.get_time_dataframe()
+        project_tz = project_time_dim.model.timezone.tz_name
+        idx = 0
+        for tz_value, tz_name in zip(geo_tz_values, geo_tz_names):
+            local_time_df = project_time_df.withColumn("time_zone", F.lit(tz_value)).withColumn(
+                "local_time",
+                F.from_utc_timestamp(F.to_utc_timestamp(F.col(ptime_col), project_tz), tz_name),
+            )
+            select = [ptime_col, "time_zone"]
+            for col in time_cols:
+                func = col.replace("_", "")
+                expr = f"{func}(local_time) AS {col}"
+                if func == "dayofweek":
+                    expr = f"{func}(local_time)-1 AS {col}"
+                select.append(expr)
+            local_time_df = local_time_df.selectExpr(*select)
+            if idx == 0:
+                time_df = local_time_df
+            else:
+                time_df = time_df.union(local_time_df)
+            idx += 1
+
+        # join all
+        join_keys = time_cols + ["time_zone"]
+        select = [
+            col if col not in time_cols else F.col(col).cast(IntegerType()) for col in df.columns
+        ]
+        df = df.select(*select).join(time_df, on=join_keys, how="left").drop(*join_keys)
+
+        # QC
+        # project_time_len = project_time_df.count()
+        # dataset_time_len = self.get_time_dataframe().count()
+        # df_len = df.count()
+        # dfn_len = dfn.count()
+        # expected_len = df_len / dataset_time_len * project_time_len
+        # assert dfn_len == expected_len, f"df with exploded time has len={dfn_len}, but expecting {expected_len}, check join logic!"
+
+        df_check = df.groupBy(ptime_col).count().sort(ptime_col)
+        freq_count = df_check.select("count").distinct().collect()
+        assert len(freq_count) == 1, freq_count
```

Comment:
This check fails on my system if the Spark Session has a Denver time zone. It passes when I set it to UTC. I'm OK if we enforce UTC, but it should be explicit (or maybe we can set it).
```
df_check.filter("count != 212").show()
+-------------------+-----+
|          timestamp|count|
+-------------------+-----+
|2012-03-10 21:00:00|  424|
+-------------------+-----+
```